File: py/thesmsworks_sdk.py
# Thesmsworks SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import ThesmsworksUtility
from core.spec import ThesmsworksSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import ThesmsworksBaseFeature
from features import _make_feature


class ThesmsworksSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = ThesmsworksUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return ThesmsworksUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = ThesmsworksSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Batch(self, data=None) -> "BatchEntity":
        """Entity factory: client.Batch().list() / client.Batch().load({"id": ...})."""
        from entity.batch_entity import BatchEntity
        return BatchEntity(self, data)


    def BatchMessage(self, data=None) -> "BatchMessageEntity":
        """Entity factory: client.BatchMessage().list() / client.BatchMessage().load({"id": ...})."""
        from entity.batch_message_entity import BatchMessageEntity
        return BatchMessageEntity(self, data)


    def Credit(self, data=None) -> "CreditEntity":
        """Entity factory: client.Credit().list() / client.Credit().load({"id": ...})."""
        from entity.credit_entity import CreditEntity
        return CreditEntity(self, data)


    def Flash(self, data=None) -> "FlashEntity":
        """Entity factory: client.Flash().list() / client.Flash().load({"id": ...})."""
        from entity.flash_entity import FlashEntity
        return FlashEntity(self, data)


    def Message(self, data=None) -> "MessageEntity":
        """Entity factory: client.Message().list() / client.Message().load({"id": ...})."""
        from entity.message_entity import MessageEntity
        return MessageEntity(self, data)


    def OneTimePassword(self, data=None) -> "OneTimePasswordEntity":
        """Entity factory: client.OneTimePassword().list() / client.OneTimePassword().load({"id": ...})."""
        from entity.one_time_password_entity import OneTimePasswordEntity
        return OneTimePasswordEntity(self, data)


    def Schedule(self, data=None) -> "ScheduleEntity":
        """Entity factory: client.Schedule().list() / client.Schedule().load({"id": ...})."""
        from entity.schedule_entity import ScheduleEntity
        return ScheduleEntity(self, data)


    def Swagger(self, data=None) -> "SwaggerEntity":
        """Entity factory: client.Swagger().list() / client.Swagger().load({"id": ...})."""
        from entity.swagger_entity import SwaggerEntity
        return SwaggerEntity(self, data)


    def Util(self, data=None) -> "UtilEntity":
        """Entity factory: client.Util().list() / client.Util().load({"id": ...})."""
        from entity.util_entity import UtilEntity
        return UtilEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "ThesmsworksSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.batch_entity import BatchEntity
    from entity.batch_message_entity import BatchMessageEntity
    from entity.credit_entity import CreditEntity
    from entity.flash_entity import FlashEntity
    from entity.message_entity import MessageEntity
    from entity.one_time_password_entity import OneTimePasswordEntity
    from entity.schedule_entity import ScheduleEntity
    from entity.swagger_entity import SwaggerEntity
    from entity.util_entity import UtilEntity

File: dart/lib/ThesmsworksSDK.dart
// Thesmsworks Dart SDK

import 'entity/BatchEntity.dart';
import 'entity/BatchMessageEntity.dart';
import 'entity/CreditEntity.dart';
import 'entity/FlashEntity.dart';
import 'entity/MessageEntity.dart';
import 'entity/OneTimePasswordEntity.dart';
import 'entity/ScheduleEntity.dart';
import 'entity/SwaggerEntity.dart';
import 'entity/UtilEntity.dart';


export 'ThesmsworksTypes.dart';
export 'entity/BatchEntity.dart';
export 'entity/BatchMessageEntity.dart';
export 'entity/CreditEntity.dart';
export 'entity/FlashEntity.dart';
export 'entity/MessageEntity.dart';
export 'entity/OneTimePasswordEntity.dart';
export 'entity/ScheduleEntity.dart';
export 'entity/SwaggerEntity.dart';
export 'entity/UtilEntity.dart';


// ignore_for_file: non_constant_identifier_names

import 'dart:async';

import 'Config.dart';
import 'Spec.dart';
import 'ThesmsworksEntityBase.dart';
import 'ThesmsworksError.dart';
import 'feature/base/BaseFeature.dart';
import 'utility/ErrUtility.dart';
import 'utility/Utility.dart';

export 'Config.dart' show Config, config;
export 'ThesmsworksEntityBase.dart' show ThesmsworksEntityBase;
export 'ThesmsworksError.dart' show ThesmsworksError;
export 'feature/base/BaseFeature.dart' show BaseFeature;
export 'utility/Utility.dart' show Utility;

final Utility stdutil = Utility();

class ThesmsworksSDK {
  String mode = 'live';
  dynamic _options;
  final Utility _utility = Utility();
  List<dynamic> features = [];
  dynamic rootctx;

  // Feature activity tracking store (retry attempts, cache hits, spans, ...).
  final Map<String, dynamic> track = {};

  ThesmsworksSDK([dynamic options]) {
    rootctx = _utility.makeContext({
      'client': this,
      'utility': _utility,
      'config': config.toMap(),
      'options': options,
      'shared': {},
    });

    _options = _utility.makeOptions(rootctx);

    final struct = _utility.struct;

    if (true == struct.getpath(_options, 'feature.test.active')) {
      mode = 'test';
    }

    rootctx.options = _options;

    features = [];

    final featureAdd = _utility.featureAdd;
    final featureInit = _utility.featureInit;

    // Add features in the resolved order (makeOptions puts an explicit List
    // order first, else defaults to test-first). Ordering matters: the
    // `test` feature installs the base mock transport and the transport
    // features (retry/cache/netsim/proxy/ratelimit) wrap whatever is current,
    // so `test` must be added before them to sit at the base of the chain.
    final featureorder =
        struct.getpath(_options, '__derived__.featureorder') ?? [];
    for (final fname in featureorder) {
      final fopts = _options['feature'][fname];
      if (fopts is Map && true == fopts['active']) {
        featureAdd(rootctx, config.makeFeature(fname.toString()));
      }
    }

    if (null != _options['extend']) {
      for (final f in _options['extend']) {
        featureAdd(rootctx, f);
      }
    }

    for (final f in features) {
      featureInit(rootctx, f);
    }

    final featureHook = _utility.featureHook;
    featureHook(rootctx, 'PostConstruct');
  }

  dynamic options() {
    return _utility.struct.clone(_options);
  }

  Utility utility() {
    return _utility;
  }

  Future<dynamic> prepare([dynamic fetchargs]) async {
    final utility = _utility;

    final makeContext = utility.makeContext;
    final makeFetchDef = utility.makeFetchDef;
    final prepareHeaders = utility.prepareHeaders;
    final prepareAuth = utility.prepareAuth;

    fetchargs = fetchargs ?? {};

    final ctx = makeContext({
      'opname': 'prepare',
      'ctrl': fetchargs['ctrl'] ?? {},
    }, rootctx);

    final options = _options;

    // Build spec directly from SDK options + user-provided fetch args.
    final spec = Spec({
      'base': options['base'],
      'prefix': options['prefix'],
      'suffix': options['suffix'],
      'path': fetchargs['path'] ?? '',
      'method': fetchargs['method'] ?? 'GET',
      'params': fetchargs['params'] ?? {},
      'query': fetchargs['query'] ?? {},
      'body': fetchargs['body'],
      'step': 'start',
    });

    ctx.spec = spec;

    spec.headers = prepareHeaders(ctx);

    // Merge user-provided headers over SDK defaults.
    if (fetchargs['headers'] is Map) {
      (fetchargs['headers'] as Map).forEach((key, val) {
        spec.headers[key] = val;
      });
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    final authResult = prepareAuth(ctx);
    if (iserr(authResult)) {
      return authResult;
    }

    return makeFetchDef(ctx);
  }

  Future<dynamic> direct([dynamic fetchargs]) async {
    final utility = _utility;
    final fetcher = utility.fetcher;
    final makeContext = utility.makeContext;

    final fetchdef = await prepare(fetchargs);
    if (iserr(fetchdef)) {
      return fetchdef;
    }

    final ctx = makeContext({
      'opname': 'direct',
      'ctrl': (fetchargs ?? {})['ctrl'] ?? {},
    }, rootctx);

    try {
      final dynamic fetched =
          await Future.value(fetcher(ctx, fetchdef['url'], fetchdef));

      if (null == fetched) {
        return {
          'ok': false,
          'err': ctx.error('direct_no_response', 'response: undefined')
        };
      } else if (iserr(fetched)) {
        return {'ok': false, 'err': fetched};
      }

      final status = fetched['status'];

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing.
      final headers = fetched['headers'];
      final contentLength =
          headers is Map ? headers['content-length'] : null;
      final noBody = 204 == status ||
          304 == status ||
          '0' == (null == contentLength ? null : contentLength.toString());

      dynamic json;
      if (!noBody) {
        try {
          final jsonFn = fetched['json'];
          json = jsonFn is Function
              ? await Future.value(jsonFn())
              : fetched['json'];
        } catch (_parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays null; callers can inspect status/headers.
          json = null;
        }
      }

      return {
        'ok': status is num && status >= 200 && status < 300,
        'status': status,
        'headers': fetched['headers'],
        'data': json,
      };
    } catch (err) {
      return {'ok': false, 'err': err};
    }
  }


  // Entity access: `client.Batch().list()` / `client.Batch().load({'id': ...})`.
  BatchEntity Batch([dynamic entopts]) {
    return BatchEntity(this, entopts);
  }


  // Entity access: `client.BatchMessage().list()` / `client.BatchMessage().load({'id': ...})`.
  BatchMessageEntity BatchMessage([dynamic entopts]) {
    return BatchMessageEntity(this, entopts);
  }


  // Entity access: `client.Credit().list()` / `client.Credit().load({'id': ...})`.
  CreditEntity Credit([dynamic entopts]) {
    return CreditEntity(this, entopts);
  }


  // Entity access: `client.Flash().list()` / `client.Flash().load({'id': ...})`.
  FlashEntity Flash([dynamic entopts]) {
    return FlashEntity(this, entopts);
  }


  // Entity access: `client.Message().list()` / `client.Message().load({'id': ...})`.
  MessageEntity Message([dynamic entopts]) {
    return MessageEntity(this, entopts);
  }


  // Entity access: `client.OneTimePassword().list()` / `client.OneTimePassword().load({'id': ...})`.
  OneTimePasswordEntity OneTimePassword([dynamic entopts]) {
    return OneTimePasswordEntity(this, entopts);
  }


  // Entity access: `client.Schedule().list()` / `client.Schedule().load({'id': ...})`.
  ScheduleEntity Schedule([dynamic entopts]) {
    return ScheduleEntity(this, entopts);
  }


  // Entity access: `client.Swagger().list()` / `client.Swagger().load({'id': ...})`.
  SwaggerEntity Swagger([dynamic entopts]) {
    return SwaggerEntity(this, entopts);
  }


  // Entity access: `client.Util().list()` / `client.Util().load({'id': ...})`.
  UtilEntity Util([dynamic entopts]) {
    return UtilEntity(this, entopts);
  }



  static ThesmsworksSDK test([dynamic testoptsarg, dynamic sdkoptsarg]) {
    final struct = stdutil.struct;
    final setpath = struct.setpath;
    final getdef = struct.getdef;
    final clone = struct.clone;
    final setprop = struct.setprop;

    final sdkopts = getdef(clone(sdkoptsarg), {});
    final testopts = getdef(clone(testoptsarg), {});
    setprop(testopts, 'active', true);
    setpath(sdkopts, 'feature.test', testopts);

    final testsdk = ThesmsworksSDK(sdkopts);
    testsdk.mode = 'test';

    return testsdk;
  }

  ThesmsworksSDK tester([dynamic testopts, dynamic sdkopts]) {
    return ThesmsworksSDK.test(testopts, sdkopts);
  }

  Map<String, dynamic> toJSON() {
    return {'name': 'Thesmsworks'};
  }

  @override
  String toString() {
    return 'Thesmsworks ' + _utility.struct.jsonify(toJSON());
  }
}

typedef SDK = ThesmsworksSDK;
